src/cargo/core/dependency.rs

```diff
@@ -103,6 +105,25 @@ impl DependencyInner {
         })
     }
 
+    fn parse_with_deprecated(req: &str, config: &Config) -> CargoResult<VersionReq> {
+        match VersionReq::parse(req) {
+            Err(e) => {
+                match e {
+                    ReqParseError::DeprecatedVersionRequirement(requirement) => {
+                        let msg = format!("One of your version requirements ({}) is invalid. \
+Previous versions of Cargo accepted this malformed requirement, but it is being deprecated. Please \
+use {} instead.", req, requirement);
```

Comment:
I think we'll emit this for *anything* cargo parses, so perhaps something like:

```
parsed version requirement `{}` is no longer valid

Previous versions of Cargo accepted this malformed requirement,
but it is being deprecated. This was found when parsing the manifest
of iron v1.2.3, and the correct version requirement is `0.2.0`.

This will soon become a hard error, so it's either recommended to
update to a fixed version or contact the upstream maintainer about
this warning.
```

That is, the requirement may not always be local, so we'll just want to help everyone find the crate at fault asap.